### ISSUE TITLE
Compose implementation outdated

### ISSUE THREAD

**Issue by LexBosch** [webdataset/webdataset#163]
I've been trying to duplicate the compose implementation given in the documentation, but copying the source gives me the following error.

> import matplotlib.pyplot as plt
> import torch.utils.data
> import torch.nn
> from random import randrange
> import os
> os.environ["WDS_VERBOSE_CACHE"] = "1"
> 
> import webdataset as wds
> 
> url = "http://storage.googleapis.com/nvdata-publaynet/publaynet-train-{000000..000009}.tar"
> dataset = wds.WebDataset(url).shuffle(1000).decode("rgb").to_tuple("png", "json")
> 
> for image, json in dataset:
>     break
> plt.imshow(image)
> 
> def preprocess(sample):
>     image, json = sample
>     try:
>         label = json["annotations"][0]["category_id"]
>     except:
>         label = 0
>     return 1-image, label
> 
> dataset = dataset.map(preprocess)
> 
> for image, label in dataset:
>     break
> plt.imshow(image)
> print(label)
> 
> 
> def get_patches(src):
>     for sample in src:
>         image, label = sample
>         h, w = image.shape[:2]
>         for i in range(16):
>             y, x = randrange(h - 256), randrange(w - 265)
>             patch = image[y:y + 256, x:x + 256]
>             yield (patch, label)
> 
> 
> dataset = dataset.compose(get_patches).shuffle(10000)

This gives

> File "~/venv/lib/python3.8/site-packages/webdataset/composable.py", line 65, in compose
>     return constructor(*args, **kw).source_(self)

Am I misunderstanding the documentation or is something out of date?

**Comment by tmbdev** [comment]
The recommended interface is the wds.DataPipeline interface. It's easier to use and easier to extend.

The .compose(...) interface is slightly different and there is no source_ anymore. Conversion should be very simple, though. Have a look at compat.py to see how the implementation changed.

**Comment by cfanyyx** [comment]
How to use SampleEqually(IterableDataset, wds.Shorthands, wds.Composable) under new API?

**Comment by cyrildiagne** [comment]
Is there any example available on how to sample equally from multiple datasets using that new interface?

**Comment by dome272** [comment]
Interested too. As a side note: Thanks for all the work on webdataset. Appreciate it!

**Comment by samar-khanna** [comment]
I used this to re-implement `SampleEqually`. It would be good to update the docs with the official solution.
```
class SampleEqually(wds.DataPipeline, wds.compat.FluidInterface):
    def __init__(self, datasets):
        super().__init__()
        self.datasets = datasets

    def __iter__(self):
        sources = [iter(ds) for ds in self.datasets]
        while True:
            for source in sources:
                try:
                    yield next(source)
                except StopIteration:
                    return
```
